base/src/com/google/idea/blaze/base/projectview/section/ListSection.java

```diff
@@ -88,16 +90,29 @@ public static <T> Builder<T> update(
   public static class Builder<T> extends SectionBuilder<T, ListSection<T>> {
     private final List<ItemOrTextBlock<T>> items = new ArrayList<>();
 
+    private int firstLineNumber = -1;
+
     public Builder(SectionKey<T, ListSection<T>> sectionKey, @Nullable ListSection<T> section) {
       super(sectionKey);
       if (section != null) {
         items.addAll(section.itemsOrComments);
       }
     }
 
+    public Builder<T> setFirstLineNumber(int firstLineNumber) {
+      this.firstLineNumber = firstLineNumber;
+      return this;
+    }
+
     @CanIgnoreReturnValue
     public final Builder<T> add(T item) {
-      items.add(new ItemOrTextBlock<>(item));
+      items.add(new ItemOrTextBlock<>(item, -1));
```

Comment:
When do we want to call this method? It feels like the only valid states an item should have are "with a valid line number" and _maybe_ "with the previous line number + 1", I'm not sure when this is a valid state.

Reply:
This one can only be used when you are creating a project view file from scratch. At that time, you don't know anything about line numbers. So where do we get the line numbers from? These add commands always have a print after them. If you look at the print commands you will see that they alter the line numbers. So this is more than a temporary state that is complemented by actually writing to the file.